paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreRead.java

```diff
@@ -267,17 +277,34 @@ private RecordReader<KeyValue> batchMergeRead(
         return keyProjectedFields == null ? reader : projectKey(reader, keyProjectedFields);
     }
 
-    private RecordReader<KeyValue> streamingConcat(
-            List<DataFileMeta> files, KeyValueFileReaderFactory readerFactory) throws IOException {
+    private RecordReader<KeyValue> noMergeRead(
+            BinaryRow partition,
+            int bucket,
+            List<DataFileMeta> files,
+            @Nullable List<DeletionFile> deletionFiles,
+            boolean onlyFilterKey)
+            throws IOException {
+        KeyValueFileReaderFactory readerFactory =
+                readerFactoryBuilder.build(
+                        partition, bucket, true, onlyFilterKey ? filtersForKeys : filtersForAll);
         List<ReaderSupplier<KeyValue>> suppliers = new ArrayList<>();
-        for (DataFileMeta file : files) {
+        for (int i = 0; i < files.size(); i++) {
+            DataFileMeta file = files.get(i);
+            DeletionFile deletionFile = deletionFiles == null ? null : deletionFiles.get(i);
             suppliers.add(
                     () -> {
                         // We need to check extraFiles to be compatible with Paimon 0.2.
                         // See comments on DataFileMeta#extraFiles.
                         String fileName = changelogFile(file).orElse(file.fileName());
-                        return readerFactory.createRecordReader(
-                                file.schemaId(), fileName, file.fileSize(), file.level());
+                        RecordReader<KeyValue> reader =
+                                readerFactory.createRecordReader(
+                                        file.schemaId(), fileName, file.fileSize(), file.level());
+                        if (deletionFile != null) {
+                            DeletionVector deletionVector =
+                                    DeletionVector.read(fileIO, deletionFile);
```

Comment:
Maybe can add this to IndexFileHandler, readDeletionVector contains version and crc verification

    public DeletionVector readDeletionVector(DeletionFile deletionFile) {
        return deletionVectorsIndex.readDeletionVector(
                new Path(deletionFile.path()),
                Pair.of(deletionFile.offset(), deletionFile.length()));
    }

Reply:
I don't want introduce `IndexFileHandler` here, imagine a C++Reader, which should not rely on any complex classes, the simpler the better.